pyro/contrib/funsor/named_messenger.py

```diff
@@ -0,0 +1,154 @@
+        level can depend on each other; if ``keep=False``, neighboring branches
+        are independent (conditioned on their shared ancestors).
+    """
+    def __init__(self, history=1, keep=False):
+        assert history >= 0
+        self.history = history
+        self.keep = keep
+        self._iterable = None
+        self._pos = -1
+        self._stack = []
+        super().__init__()
+
+    def generator(self, iterable):
+        self._iterable = iterable
+        return self
+
+    def __iter__(self):
+        with ExitStack() as stack:
+            for value in self._iterable:
+                stack.enter_context(self)
+                yield value
+
+    def __enter__(self):
+        self._pos += 1
+        if len(self._stack) <= self._pos:
+            self._stack.append(
+                StackFrame(name_to_dim=OrderedDict(), dim_to_name=OrderedDict()))
+        return super().__enter__()
+
+    def __exit__(self, *args, **kwargs):
+        if not self.keep:
+            self._stack.pop()
+        self._pos -= 1
+        return super().__exit__(*args, **kwargs)
+
+    @property
+    def current_frame(self):
+        return self._stack[self._pos]
+
+    def _pyro_to_data(self, msg):
+        if len(msg["args"]) < 2:
+            msg["args"] = msg["args"] + (OrderedDict(),)
+        funsor_value, name_to_dim = msg["args"]
+
+        # every markov context is responsible for knowing about non-fresh local vars,
+        # and fresh vars are handled by passing information up & down the stack
+
+        # handling non-fresh vars: just look at the deepest context they appear in
+        name_to_dim.update({name: self.current_frame.name_to_dim[name]
+                            for name in funsor_value.inputs
+                            if name in self.current_frame.name_to_dim
+                            and name not in name_to_dim})
+
+        # handling fresh vars: pass responsibility for allocation up the stack
+        depth = msg.setdefault("depth", -self.history - 1)
+        if depth == -self.history - 1:
+            msg["fresh"] = frozenset(name for name in funsor_value.inputs
+                                     if name not in name_to_dim
+                                     and name not in self.current_frame.name_to_dim)
+        fresh = msg["fresh"]
+
+        for pos in reversed(range(max(0, self._pos - self.history), self._pos + 1)):
+            frame = self._stack[pos]
+            msg["depth"] += 1
+            if msg["depth"] == 0:
+                for name in fresh:
+                    if name not in frame.name_to_dim:
+                        dim = -1
+                        while dim in frame.dim_to_name:
+                            dim -= 1
+                        frame.name_to_dim[name] = dim
+                    dim = frame.name_to_dim[name]
+                    frame.dim_to_name[dim] = (name, funsor_value.inputs[name])
+                    name_to_dim[name] = dim
```

Comment:
`NamedMessenger._pyro_to_data` allocates fresh dimensions to variable names it hasn't seen before